clang/test/AST/ast-print-language-linkage.cpp

```diff
@@ -0,0 +1,27 @@
+// RUN: %clang_cc1 -ast-print %s -o - | FileCheck %s
+
+// CHECK: extern "C" int printf(const char *, ...);
+extern "C" int printf(const char *...);
+
+// CHECK: extern "C++" {
```

Comment:
@AaronBallman : What do we think of this?  This isn't a true representation of the original code, but I'm not sure what work we have to ensure this is done with commas?

Reply:
Unfortunately, I'm not aware of the design goals of `DeclPrinter`. How accurately is it trying to represent the original code? Because AST's are naturally a lossy projection of CST's, so there is always going to be a difference to the original. How much of that difference is admissible?

I can try to work around that by querying `VarDecl::isInExternCContext`/`isInExternCXXContext` inside of `VarDecl`, but my initial naive implementation didn't have the desired result.

Reply:
We don't really have much guarantees besides 'best effort', so I am querying Aaron to see if this is 'good enough'.  Just `VarDecl` isn't going to be sufficient (probably something in `DeclStmt` handling?) because we can have any sorts of declarations this way.

Reply:
> Oofda, there are differences between the braced and non-braced versions, so ideally, we want to reproduce exactly what was in the user's source to avoid accidentally changing the meaning of code

You are completely right, my bad. I'll work on fixing this. 

> z, y, and z are external, a, b, and c are not: https://godbolt.org/z/evG9KPKMM

I believe, they all have external linkage. However, x, y and z are also definitions, while a, b, and c are just declarations. So, for example, you can reference `x` from another translation unit and the value will be `0`.

Reply:
> ideally, we want to reproduce exactly what was in the user's source to avoid accidentally changing the meaning of code

Do you know why we split declarators from the same declaration into several declarations, whenever possible? Judging by the comments in `DeclPrinter::visitDeclContext` it seems intentional and desired.

Reply:
Hopefully, the new implementation has fixed the issue.